networkx/algorithms/summarization.py

```diff
@@ -0,0 +1,214 @@
+original graph.  The two-part minimum description length (MDL) is often used to
+represent the model and the original graph in terms of the model.  A key
+difference between graph compression and graph summarization is that graph
+summarization focuses on finding structural patterns within the original graph,
+whereas graph compression focuses on compressions the original graph to be as
+small as possible.  **NOTE**: Some bit-compression methods exist solely to
+compress a graph without creating a summary graph or finding comprehensible
+structural patterns.
+
+Simplification/Sparsification techniques attempt to create a sparse
+representation of a graph by removing unimportant nodes and edges from the
+graph.  Sparsified graphs differ from supergraphs created by
+grouping/aggregation by only containing a subset of the original nodes and
+edges of the original graph.
+
+Influence based techniques aim to find a high-level description of influence
+propagation in a large graph.  These methods are scarce and have been mostly
+applied to social graphs.
+
+*dedensification* is a grouping/aggregation based technique to compress the
+neighborhoods around high-degree nodes in unweighted graphs by adding
+compressor nodes that summarize multiple edges of the same type to
+high-degree nodes (nodes with a degree greater than a given threshold).
+Dedensification was developed for the purpose of increasing performance of
+query processing around high-degree nodes in graph databases and enables direct
+operations on the compressed graph.  The structural patterns surrounding
+high-degree nodes in the original is preserved while using fewer edges and
+adding a small number of compressor nodes.  The degree of nodes present in the
+original graph is also preserved. The current implementation of dedensification
+supports graphs with one edge type.
+
+For more information on graph summarization, see `Graph Summarization Methods
+and Applications: A Survey <https://dl.acm.org/doi/abs/10.1145/3186727>`_
+"""
+import networkx as nx
+
+__all__ = [
+    "dedensify",
+]
+
+
+def dedensify(G, threshold=2, prefix=None, inplace=False):
+    """Compresses neighborhoods around high-degree nodes
+
+    Reduces the number of edges to high-degree nodes by adding compressor nodes
+    that summarize multiple edges of the same type to high-degree nodes (nodes
+    with a degree greater than a given threshold).  Dedensification also has
+    the added benefit of reducing the number of edges around high-degree nodes.
+    The implementation currently supports graphs with a single edge type.
+
+    Parameters
+    ----------
+    G: graph
+       A networkx graph
+    threshold: int, optional (default=2)
+       Minimal number of degrees to be considered a high degree node
+    prefix: str or None, optional (default: None)
+       An optional prefix for denoting compressor nodes
+    inplace: bool, optional (default: False)
+       Indicates if dedensification should be done inplace
```

Comment:
This appears to be the only function where `inplace` is used as a kwarg. The releases notes from `api_0.99.rst`  state:

 > The argument inplace=False|True has been replaced with copy=True|False.

So maybe it'd be better to use a copy kwarg here instead?

Reply:
@rossbar Sorry about that.  I just changed the kwarg to be `copy=False`.

Reply:
The only remaining issue with this is what the default value for `copy=` should be. On the one hand, a default of `copy=True` is more in keeping with the functional-style that inputs aren't modified by the functions, but OTOH this function is likely to be used primarily on large graphs, where there's a good reason (memory usage) not to produce copies.

The results of a quick grep make it seem like the default for all other functions that have a `copy=` kwarg is `copy=True`, so I tend to lean that way. Thoughts @jarrodmillman @dschult ?

Reply:
I think it produces fewer "surprises" to have the default be `copy=True`.
When people are trying it on smaller cases to get used to it this default makes sense. When they get serious and want to use it for a big graph and save memory, they should know enough by then to be able to turn it off.
But maybe I am naive. :}

Reply:
Sorry about that...again.  I just updated it to be `copy=True`.